### PR TITLE
Return different process exit codes from iotedged after configured retry count has been exceeded

### PR DESCRIPTION
**Pull request by shantanu1singh**
Return different process exit codes from iotedged after configured retry count has been exceeded

### CODE REVIEW

File: edgelet/edgelet-core/src/watchdog.rs [hunk 1]
```diff
@@ -110,22 +116,41 @@ where
...
             .or_else(|e| {
                 warn!("Error in watchdog when checking for edge runtime status:");
                 log_failure(Level::Warn, &e);
-                future::ok(())
+                future::ok(Some(e))
             })
         })
+        .fold(
+            0,
+            move |exec_count: u32, result: Option<Error>| match result {
```
Comment (myagley) [inline edgelet/edgelet-core/src/watchdog.rs]: with the retry limit enum you can do:

```rust
move |exec_count: u32, result: Option<Error>| {
    result.and_then(|e| {
        if retry_count.cmp(exec_count) == Ordering::Greater {
            Ok(exec_count + 1)
        } else {
            Err(e)
        }
    })
    .unwrap_or_else(|| Ok(0))
}
```

Using a `match` with `Option` is generally seen as an anti-pattern. Especially, when there are only two branches, a `Some` and `None`. People tend to use `map`/`and_then` paired with one of the `unwrap_` methods.

File: edgelet/iotedged/src/error.rs [hunk 2]
```diff
@@ -25,6 +31,15 @@ pub enum ErrorKind {
...
 
+    #[fail(display = "Invalid device configuration was provided.")]
+    InvalidDeviceConfig,
+
+    #[fail(display = "Invalid IoT hub configuration was provided.")]
+    InvalidHubConfig,
+
+    #[fail(display = "Invalid signed token was provided.")]
+    InvalidSignedToken,
+
```
Comment (myagley) [inline edgelet/iotedged/src/error.rs]: We should consider moving these into `InitializeErrorReason`s or create a `Watchdog(WatchdogErrorReason)` variant, instead of adding these to the flat namespace.

Comment (myagley) [reply edgelet/iotedged/src/error.rs]: @arsing do you have any comments on the error handling?

File: edgelet/contrib/config/linux/config.yaml [hunk 3]
```diff
@@ -144,6 +144,27 @@ agent:
...
+# On a fatal failure, the daemon returns an exit code which signifies the kind
+# of error encountered. Currently, the following error codes are returned by 
+# the daemon -
+# 150 - Invalid Device ID specified.
+# 151 - Invalid IoT hub configuration.
+# 152 - Invalid SAS token used to call IoT hub. This could signal an invalid SAS key.
+# 1 - All other errors.
+###############################################################################
+
+# retry_count: 2
```
[the remark this answers is not in the document]
Comment (shantanu1singh) [reply edgelet/contrib/config/linux/config.yaml]: Sure, changing it

File: edgelet/edgelet-config/src/lib.rs [hunk 4]
```diff
@@ -368,6 +368,7 @@ pub struct Settings<T> {
     homedir: PathBuf,
     moby_runtime: MobyRuntime,
     certificates: Option<Certificates>,
+    retry_count: Option<u32>,
```
[the remark this answers is not in the document]
Comment (shantanu1singh) [reply edgelet/edgelet-config/src/lib.rs]: Sure, changing it

File: edgelet/edgelet-config/src/lib.rs [hunk 5]
```diff
@@ -432,6 +433,10 @@ where
         self.certificates.as_ref()
     }
 
+    pub fn retry_count(&self) -> Option<&u32> {
```
[the remark this answers is not in the document]
Comment (shantanu1singh) [reply edgelet/edgelet-config/src/lib.rs]: Thanks for the tip, will change it

File: edgelet/iotedged/src/error.rs (hunk 2 above)
Comment (shantanu1singh) [reply edgelet/iotedged/src/error.rs]: I agree for the hub and device config but the signed tokens can be invalid after successful initialization too i.e., the daemon can get 401 errors from IoT hub after initialization too. Should we keep that error type in the flat namespace then?

File: edgelet/edgelet-config/Cargo.toml [hunk 6]
```diff
@@ -7,7 +7,7 @@ edition = "2018"
 
 [dependencies]
 base64 = "0.9"
-config = "0.8"
+config = "0.9.2"
```
Comment (myagley) [inline edgelet/edgelet-config/Cargo.toml]: What is this change required for?

File: edgelet/iotedged/src/error.rs [hunk 7]
```diff
@@ -144,6 +217,14 @@ impl fmt::Display for InitializeErrorReason {
 
             InitializeErrorReason::HttpClient => write!(f, "Could not initialize HTTP client"),
 
+            InitializeErrorReason::InvalidDeviceConfig => {
+                write!(f, "Invalid device configuration was provide")
```
Comment (myagley) [inline edgelet/iotedged/src/error.rs]:
```suggestion
                write!(f, "Invalid device configuration was provided")
```

File: edgelet/edgelet-config/Cargo.toml (hunk 6 above)
Comment (shantanu1singh) [reply edgelet/edgelet-config/Cargo.toml]: I believe I was running into some issues earlier with the config crate version 0.8 when I was using the Option<u32> change. Have reverted this now given how the new method seems to work with 0.8 too

File: edgelet/edgelet-core/src/watchdog.rs [hunk 8]
```diff
@@ -26,9 +28,32 @@ const MODULE_GENERATIONID: &str = "IOTEDGE_MODULEGENERATIONID";
 /// This is the frequency with which the watchdog checks for the status of the edge runtime module.
 const WATCHDOG_FREQUENCY_SECS: u64 = 60;
 
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(untagged)]
+pub enum RetryLimit {
+    Infinite,
+    Num(u32),
```
Comment (arsing) [inline edgelet/edgelet-core/src/watchdog.rs]: It's not going to matter, but countable things are usually measured in `usize` rather than `u32`.

File: edgelet/iotedged/src/error.rs [hunk 9]
```diff
@@ -72,12 +81,74 @@ impl From<ErrorKind> for Error {
     }
 }
 
+impl From<CoreError> for Error {
+    fn from(error: CoreError) -> Self {
+        let fail: &dyn Fail = &error;
+        let mut error_kind = ErrorKind::Watchdog;
+
+        for cause in fail.iter_causes() {
```
[the remark this answers is not in the document]
Comment (shantanu1singh) [reply edgelet/iotedged/src/error.rs]: yeah gonna keep it this way
